core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java

```diff
@@ -0,0 +1,130 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.rest;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * {@code DeprecationRestChannel} is a proxy for an existing {@link RestChannel} that decorates the {@link RestResponse} with
+ * the {@link #DEPRECATION_HEADER deprecation header} and a warning message to denote that the existing request/response is
+ * deprecated.
+ */
+public class DeprecationRestChannel implements RestChannel {
+    /**
+     * The "Warning" Header comes from RFC-7234. As the RFC describes, it's generally used for caching purposes, but it can be
+     * used for <em>any</em> warning.
+     *
+     * https://tools.ietf.org/html/rfc7234#section-5.5
+     */
+    public static final String DEPRECATION_HEADER = "Warning";
+
+    /**
+     * The proxied canchannel.
+     */
+    private final RestChannel channel;
+    /**
+     * The deprecation message.
+     */
+    private final String warning;
+
+    /**
+     * Create a new {@link DeprecationRestChannel}.
+     *
+     * @param channel The channel to proxy
+     * @param warning The warning to add to the response. This should only be ASCII as it goes into a header!
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public DeprecationRestChannel(RestChannel channel, String warning) {
+        assert Strings.validHeaderValue(warning);
+
+        // required
+        this.channel = Objects.requireNonNull(channel);
+        this.warning = Objects.requireNonNull(warning);
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Adds the {@link #DEPRECATION_HEADER} to the {@code response}.
+     *
+     * @param response The response to send to the client
+     */
+    @Override
+    public void sendResponse(RestResponse response) {
+        response.addHeader(DEPRECATION_HEADER, warning);
+
+        // actually send the response
+        channel.sendResponse(response);
+    }
+
+    /**
+     * {@inheritDoc}
```

Comment:
> I am aware of its implicit nature, but explicit documentation shows intent while implicit documentation can mean anything. 

I'm not sure what you mean by this? The Javadoc for this method will have exactly the same documentation whether or not the `inheritdoc` tag is there.

> I would strongly prefer to keep these here even though I know that it's implicit.

It just adds source code clutter; I think that it should be removed. This is also inconsistent with how we do intend to do it everywhere else in the codebase (is it 100% perfect? probably not